File: apps/microkernel/init/src/http_server.rs
//! Simple HTTP server.
//!
//! Benchmark with [Apache HTTP server benchmarking tool](https://httpd.apache.org/docs/2.4/programs/ab.html):
//!
//! ```
//! ab -n 5000 -c 20 http://X.X.X.X:5555/
//! ```

extern crate alloc;

use libax::{axerrno::AxResult, io::File, task::yield_now};

const LOCAL_IP: &str = "10.0.2.15";
const LOCAL_PORT: u16 = 5555;

macro_rules! header {
    () => {
        "\
HTTP/1.1 200 OK\r\n\
Content-Type: text/html\r\n\
Content-Length: {}\r\n\
Connection: close\r\n\
\r\n\
{}"
    };
}

const CONTENT: &str = r#"<html>
<head>
  <title>Hello, ArceOS</title>
</head>
<body>
  <center>
    <h1>Hello, <a href="https://github.com/rcore-os/arceos">ArceOS</a></h1>
  </center>
  <hr>
  <center>
    <i>Powered by <a href="https://github.com/rcore-os/arceos/tree/main/apps/net/httpserver">ArceOS example HTTP server</a> v0.1.0</i>
  </center>
</body>
</html>
"#;

fn http_server(stream: File) -> AxResult {
    let mut buf = [0u8; 1024];
    stream.read(&mut buf)?;

    let reponse = alloc::format!(header!(), CONTENT.len(), CONTENT);
    stream.write(reponse.as_bytes())?;

    Ok(())
}

fn accept_loop() -> AxResult {
    let (addr, port) = (LOCAL_IP, LOCAL_PORT);
    let path = &alloc::format!("tcp:/{}/{}", addr, port);
    println!("{}", path);
    let listener = File::create(path).unwrap();
    println!("listen on: {}", path);

    let mut i = 0;
    loop {
        match listener.dup("accept") {
            Ok(stream) => {
                info!("new client {}: ", i);
                libax::task::spawn(move || match http_server(stream) {
                    Err(e) => error!("client connection error: {:?}", e),
                    Ok(()) => info!("client {} closed successfully", i),
                });
            }
            Err(libax::axerrno::AxError::Again) => yield_now(),
            Err(e) => return Err(e),
        }
        i += 1;
    }
}

pub fn main() {
    println!("Hello, ArceOS HTTP server!");
    accept_loop().expect("test HTTP server failed");
}

File: apps/c/httpserver/httpserver.c

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>

const char header[] = "\
HTTP/1.1 200 OK\r\n\
Content-Type: text/html\r\n\
Content-Length: %u\r\n\
Connection: close\r\n\
\r\n\
";

const char content[] = "<html>\n\
<head>\n\
  <title>Hello, ArceOS</title>\n\
</head>\n\
<body>\n\
  <center>\n\
    <h1>Hello, <a href=\"https://github.com/rcore-os/arceos\">ArceOS</a></h1>\n\
  </center>\n\
  <hr>\n\
  <center>\n\
    <i>Powered by <a href=\"https://github.com/rcore-os/arceos/tree/main/apps/net/httpserver\">ArceOS example HTTP server</a> v0.1.0</i>\n\
  </center>\n\
</body>\n\
</html>\n\
";

int main()
{
    puts("Hello, ArceOS C HTTP server!");
    struct sockaddr_in local, remote;
    int addr_len = sizeof(remote);
    local.sin_family = AF_INET;
    if (inet_pton(AF_INET, "0.0.0.0", &(local.sin_addr)) != 1) {
        perror("inet_pton() error");
        return -1;
    }
    local.sin_port = htons(5555);
    int sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (sock == -1) {
        perror("socket() error");
        return -1;
    }
    if (bind(sock, (struct sockaddr *)&local, sizeof(local)) != 0) {
        perror("bind() error");
        return -1;
    }
    if (listen(sock, 0) != 0) {
        perror("listen() error");
        return -1;
    }
    puts("listen on: http://0.0.0.0:5555/");
    char buf[1024] = {};
    int client;
    char response[1024] = {};
    snprintf(response, 1024, header, strlen(content));
    strcat(response, content);
    for (;;) {
        client = accept(sock, (struct sockaddr *)&remote, (socklen_t *)&addr_len);
        if (client == -1) {
            perror("accept() error");
            return -1;
        }
        printf("new client %d\n", client);
        if (recv(client, buf, 1024, 0) == -1) {
            perror("recv() error");
            return -1;
        }
        ssize_t l = send(client, response, strlen(response), 0);
        if (l == -1) {
            perror("send() error");
            return -1;
        }
        if (close(client) == -1) {
            perror("close() error");
            return -1;
        }
        printf("client %d close: %ld bytes sent\n", client, l);
    }
    if (close(sock) == -1) {
        perror("close() error");
        return -1;
    }
    return 0;
}
